PsfLauncher/PsfPowershellScriptRunner.h

```diff
@@ -0,0 +1,305 @@
+		}
+
+		std::wstring commandString = MakeCommandString(*startScriptInformation, scriptPath);
+		DWORD scriptTimeout = GetTimeout(*startScriptInformation);
+
+
+		int showWindowAction = GetShowWindowAction(*startScriptInformation);
+		bool runInVirtualEnvironment = GetRunInVirtualEnvironment(*startScriptInformation);
+
+		if (waitForScriptToFinish)
+		{
+			HRESULT startScriptResult = StartProcess(nullptr, commandString.data(), currentDirectory.c_str(), showWindowAction, runInVirtualEnvironment, scriptTimeout);
+
+			if (stopOnScriptError)
+			{
+				THROW_IF_FAILED(startScriptResult);
+			}
+		}
+		else
+		{
+			//We don't want to stop on an error and we want to run async
+			std::thread(StartProcess, nullptr, commandString.data(), currentDirectory.c_str(), showWindowAction, runInVirtualEnvironment, scriptTimeout);
+		}
+
+		return {};
+	}
+
+	void RunEndingScript()
+	{
+		if (!endScriptInformation)
+		{
+			return;
+		}
+
+		const std::wstring& scriptPath = GetScriptPath(*endScriptInformation);
+
+		//This method throws an exception if the script does not exist
+		ThrowIfScriptDoesNotExists(scriptPath);
+
+		bool canScriptRun = false;
+		THROW_IF_FAILED(CheckIfShouldRun(canScriptRun, *endScriptInformation));
+
+		if (canScriptRun)
+		{
+			std::wstring commandString = MakeCommandString(*endScriptInformation, scriptPath);
+			DWORD scriptTimeout = GetTimeout(*endScriptInformation);
+
+
+			int showWindowAction = GetShowWindowAction(*endScriptInformation);
+			bool runInVirtualEnvironment = GetRunInVirtualEnvironment(*endScriptInformation);
+
+			THROW_IF_FAILED(StartProcess(nullptr, commandString.data(), currentDirectory.c_str(), showWindowAction, runInVirtualEnvironment, scriptTimeout));
+		}
+	}
+
+private:
+
+	std::unique_ptr<const psf::json_object> startScriptInformation;
+	std::unique_ptr<const psf::json_object> endScriptInformation;
+	std::filesystem::path currentDirectory;
+	bool stopOnScriptError = false;
+
+	std::wstring MakeCommandString(const psf::json_object& scriptInformation, const std::wstring& scriptPath) noexcept
+	{
+		std::wstring commandString = L"Powershell.exe -file StartingScriptWrapper.ps1 ";
+		commandString.append(L"\".\\");
+		commandString.append(scriptPath);
+
+		//Script arguments are optional.
+		auto scriptArgumentsJObject = scriptInformation.try_get("scriptArguments");
+		if (scriptArgumentsJObject)
+		{
+			commandString.append(L" ");
+			commandString.append(scriptArgumentsJObject->as_string().wide());
+		}
+
+		//Add ending quote for the script inside a string literal.
+		commandString.append(L"\"");
+
+		return commandString;
+	}
+
+	const std::wstring GetScriptPath(const psf::json_object& scriptInformation) const
+	{
+		//Get throws if the key does not exist.
+		return scriptInformation.get("scriptPath").as_string().wide();
+	}
+
+	DWORD GetTimeout(const psf::json_object& scriptInformation)
+	{
+		DWORD timeout = INFINITE;

```

Comment:
You can use the same pattern as GetShowWindowAction to reduce code.

Reply:
Something like this?

---
In reply to: [316822950](https://github.com/microsoft/MSIX-PackageSupportFramework/pull/79#discussion_r316822950) [](ancestors = 316822950)

Reply:
Looks good!

---
In reply to: [316937041](https://github.com/microsoft/MSIX-PackageSupportFramework/pull/79#discussion_r316937041) [](ancestors = 316937041,316822950)